client-go/rest: finish context support

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Some of the APIs in client-go/rest ignored the callers context:
- BackoffManager ignored timeout and cancellation. This is reason enough to formally deprecate it.
- Some calls logged using the global logger. This only relevant for code which cares about contextual logging, so the approach from https://github.com/kubernetes/kubernetes/issues/126379 is used to warn (eventually) about it such code and not in others.

#### Which issue(s) this PR fixes:
Related-to https://github.com/kubernetes/kubernetes/issues/126379

#### Special notes for your reviewer:

Current controller-runtime compiles okay with these changes.

This completes the work started in https://github.com/kubernetes/kubernetes/pull/126999.

#### Does this PR introduce a user-facing change?
```release-note
client-go/rest: fully supports contextual logging. BackoffManagerWithContext should be used instead of BackoffManager to ensure that the caller can interrupt the sleep.
```

/wg structured-logging
/sig instrumentation
/sig api-machinery

/cc @sttts

pohly committed:
client-go/test: warning handler with contextual logging

The default handler now uses contextual logging. Instead of

     warnings.go:106] warning 1

it now logs the caller of client-go and uses structured, contextual
logging

     main.go:100] "Warning" message="warning 1"

Users of client-go have the choice whether the handler that they provide uses
the traditional API (no API break!) or contextual logging.

pohly committed:
client-go/rest: finish conversion to contextual logging

The remaining calls can be converted without API changes.

pohly committed:
client-go/rest: backoff with context support

The BackoffManager interface sleeps without considering the caller's context,
i.e. cancellation is not supported. This alone is reason enough to deprecate it
and to replace it with an interface that supports a context parameter.

The other reason is that contextual logging needs that parameter.

pohly commented:
/test pull-kubernetes-apidiff

Do we care enough about "controller-runtime continues to build" that we should try that in the `pull-kubernetes-apidiff` job? It's advisory, so it would be okay if it fails to build when we decide to make a breaking change.

We could do similar testing with other downstream consumers, but breaking controller-runtime is probably worse because it affects all controllers built on top of it.

/cc @sbueringer

sbueringer commented:
> /test pull-kubernetes-apidiff
> 
> Do we care enough about "controller-runtime continues to build" that we should try that in the `pull-kubernetes-apidiff` job? It's advisory, so it would be okay if it fails to build when we decide to make a breaking change.
> 
> We could do similar testing with other downstream consumers, but breaking controller-runtime is probably worse because it affects all controllers built on top of it.
> 
> /cc @sbueringer

I think it could be an interesting data point

(cc @alvaroaleman)

aojea commented:
> We could do similar testing with other downstream consumers, but breaking controller-runtime is probably worse because it affects all controllers built on top of it.

There are a lot of controllers that build on top of client-go, don't know exactly the population but is not negligible 

/test pull-kubernetes-apidiff

sbueringer commented:
> > We could do similar testing with other downstream consumers, but breaking controller-runtime is probably worse because it affects all controllers built on top of it.
> 
> There are a lot of controllers that build on top of client-go, don't know exactly the population but is not negligible
> 

Definitely! Having CR as one data point could be valuable as it's widely used (https://github.com/search?q=controller-runtime++path%3Ago.mod&type=code, https://github.com/search?q=controller-runtime++path%3Ago.mod&type=code), but of course we stil have to keep folks in mind that are using client-go more directly

pohly commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,31 +51,61 @@ var (
 // WarningLogger is an implementation of WarningHandler that logs code 299 warnings
 type WarningLogger struct{}
 
-func (WarningLogger) HandleWarningHeader(code int, agent string, message string) {
```
There also was no reason to add the separate types for the `WithContext` methods. The same type can provide the old and new API.

pohly replied:
Pushed, let's see...

/test pull-kubernetes-apidiff

pohly replied:
Looks good, only false positives about unrelated types (an issue with apidiff's handling of generics) and:
```
 Compatible changes:
- ./rest.(*NoBackoff).CalculateBackoffWithContext: added
- ./rest.(*NoBackoff).SleepWithContext: added
- ./rest.(*NoBackoff).UpdateBackoffWithContext: added
- ./rest.(*Request).BackOffWithContext: added
- ./rest.(*Request).WarningHandlerWithContext: added
- ./rest.(*URLBackoff).CalculateBackoffWithContext: added
- ./rest.(*URLBackoff).SleepWithContext: added
- ./rest.(*URLBackoff).UpdateBackoffWithContext: added
- ./rest.BackoffManagerWithContext: added
- ./rest.Config.WarningHandlerWithContext: added
- ./rest.NoWarnings.HandleWarningHeaderWithContext: added
- ./rest.SetDefaultWarningHandlerWithContext: added
- ./rest.WarningHandlerWithContext: added
- ./rest.WarningLogger.HandleWarningHeaderWithContext: added 
```

seans3 commented:
/triage accepted

pohly commented:
/test pull-kubernetes-apidiff

Now with trial build against controller-runtime :smile:

aojea commented:
> /test pull-kubernetes-apidiff
> 
> Now with trial build against controller-runtime 😄

it is failing, isn't it? I mean, isnot working

pohly commented:
"main" != "master" ... => https://github.com/kubernetes/test-infra/pull/33585

aojea commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-apidiff

pohly commented:
/test pull-kubernetes-apidiff 

The false positives in the upstream apidiff might be fixed now.

pohly commented:
Do we want `pull-kubernetes-apidiff` if there are any changes detected by apidiff, even if they are compatible (current behavior) or do we want it to pass in that case? I'm leaning towards keeping the current behavior because API additions might also need a review.

pohly commented:
/retest

sttts commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,33 +51,70 @@ var (
+//
+// logcheck:context // SetDefaultWarningHandlerWithContext should be used instead of SetDefaultWarningHandler in code which supports contextual logging.
 func SetDefaultWarningHandler(l WarningHandler) {
+	defaultWarningHandlerLock.Lock()
+	defer defaultWarningHandlerLock.Unlock()
+	defaultWarningHandler = warningLoggerNopContext{l: l}
```
Nit:
```suggestion
	SetDefaulthandlerWithContext(warningLoggerNopContext{l: l})
```

sttts replied:
There are some more below that look similar.

pohly replied:
I'll fix this tomorrow, then @aojea and/or @sbueringer can re-add the LGTM.

pohly replied:
Fixed. The one below in `getDefaultWarningHandler` (if that's the one) is different.

sttts commented:
Made it through. Only small nit.

/lgtm
/approve
/hold

@aojea @sbueringer hold for you for a last look. Please unhold when you both are ok to move forward.

aojea commented:
/test pull-kubernetes-apidiff

sbueringer commented:
All good from my side

/lgtm

aojea commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -1517,32 +1518,48 @@ func TestBackoffLifecycle(t *testing.T) {
 		)}
 
 	for _, sec := range seconds {
-		thisBackoff := request.backoff.CalculateBackoff(request.URL())
```
if we have to Backoff implementation it seems better to keep the existing testing and a add a new one for the new behavior

pohly replied:
`request.backoff` is now of type `BackoffManagerWithContext`, so the old test no longer works and has to be updated.

pohly replied:
Coverage before:
```console
$ go test -cover ./staging/src/k8s.io/client-go/rest
ok  	k8s.io/client-go/rest	15.260s	coverage: 79.1% of statements
```

After (= e15d5b9da913a2e7e7d190a8aa1d6c577872f015):
```
ok  	k8s.io/client-go/rest	144.304s	coverage: 78.0% of statements
```

It has dropped a bit because some of the "legacy" wrappers are not called. Copying entire tests just to ensure that the "old" function still works IMHO is overkill because they exercise the "old", unchanged behavior. Instead, I'll add some test specifically for new code.

This and https://github.com/kubernetes/kubernetes/pull/127709#pullrequestreview-2402201817 are going to take a bit longer than we have before code freeze. Let's not rush it and target 1.33.

pohly replied:
With those additional tests we are now at 79.6%, better than before (for example, setting warning handlers was not covered). I also found one small issue, which reinforces the value of unit testing: setting nil as WarningHandler got nil wrapped instead of unsetting the handler.

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -628,20 +628,26 @@ func (r *Request) tryThrottleWithInfo(ctx context.Context, retryInfo string) err
 
 	var message string
+	var kv []any
+	if retryInfo == "" {
+		retryInfo = "client-side throttling, not priority and fairness"
+	}
```
I think I meant to replace that switch because I wasn't happy with that approach, but didn't finish. Now it's simpler:

```Go
	if latency > longThrottleLatency {
		if retryInfo == "" {
			retryInfo = "client-side throttling, not priority and fairness"
		}
		klog.FromContext(ctx).V(3).Info("Waited before sending request", "delay", latency, "reason", retryInfo, "verb", r.verb, "URL", r.URL()}
	}
```

pohly replied:
Pushed: https://github.com/kubernetes/kubernetes/compare/f97e9e1d83bcdc87a1d6ef0766453975a43a35ba..6456dd6c73c3bbe15c1c8cdf85054908f212962e

pohly commented:
/test pull-kubernetes-e2e-gce

"Services should implement NodePort and HealthCheckNodePort correctly when ExternalTrafficPolicy changes" ran into a timeout ("expected response containing : context deadline exceeded").

soltysh reviewed (approved):
/lgtm
@aojea any additional comments?

pohly commented:
Let's also hold for a decision on https://github.com/kubernetes/kubernetes/pull/126387/files#r1810731571

pohly commented:
> Let's also hold for a decision on https://github.com/kubernetes/kubernetes/pull/126387/files#r1810731571

This PR doesn't use "TODO logcheck:context", so it's fine.

As Jordan had comments on https://github.com/kubernetes/kubernetes/pull/126387 about proper context support, I did another sweep over this PR. The only potential problem that I spotted is around `loggingCtx`: if, for whatever reason, that context should get passed to function which honors cancellation (not the case right now!), then that function might get canceled unintentionally.

To prevent this potential problem, I [added](https://github.com/kubernetes/kubernetes/compare/6456dd6c73c3bbe15c1c8cdf85054908f212962e..1b263b9f35af158bd204af3562b74f483d96b7d3) `context.WithoutCancel` in all places that set that `loggingCtx`. It's [cheap](https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/context/context.go;l=568-600;drc=70e453b4361b80a85e6ebb37e4d43ec02db9a50a), so there's no drawback.

I now need to rebase...

soltysh reviewed (commented):
Left a few comments.

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -1523,6 +1546,10 @@ type Result struct {
 
+	// Log calls in Result methods use the same context for logging as the
+	// method which created the Result. This context has no cancellation.
+	loggingCtx context.Context
```
The `loggingCtx` is needed for:
```
	out, _, err := r.decoder.Decode(r.body, &schema.GroupVersionKind{Version: "v1"}, nil)
	if err != nil {
		ctx := r.loggingCtx
		if ctx == nil {
			ctx = context.Background()
		}
		klog.FromContext(ctx).V(5).Info("Body was not decodable (unable to check for Status)", "err", err)
		return r.err
	}
```

I agree that this isn't particularly likely, but consider if we don't convert this and you are trying to debug an issue that occurs for one particular request by grepping for some unique key/value pair that was set for all activities related to that request: this log entry here might explain the issue, but you won't see it.

I could extract the logger unconditionally when creating the request. That then incurs the overhead for `FromContext` whether this error occurs or not. OTOH, `WithoutCancel` also isn't free. Without good benchmarks it's hard to tell which one is cheaper, and both are probably cheap compared to the cost of handling a request.

Yet another alternative is to store a `getLogger` function in the `Request` instead of the context. Instead of `WithoutCancel`, the field would get initialized with `getLogger: func() klog.Logger { klog.FromContext(ctx} }`. But that then binds the context for longer.

I'm leaning towards storing the logger - changed in a separate commit for review and to document the decision.

pohly committed:
client-go rest: store logger in Result

Storing a context and making sure that it never gets canceled also has
overhead. We might as well just do the klog.FromContext when constructing
the Result and store the logger for later use.

pohly commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -126,10 +126,23 @@ type Config struct {
 	WarningHandler WarningHandler
 
+	// WarningHandlerWithContext handles warnings in server responses.
+	// If this and WarningHandlerWithContext are not set, the
```
Fixed, thanks.

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -260,20 +260,43 @@ func (r *Request) Resource(resource string) *Request {
+// If set to nil, this client will use the default warning handler (see [SetDefaultWarningHandler]).
+//
+//logcheck:context // WarningHandlerWithContext should be used instead of WarningHandler in code which supports contextual logging.
 func (r *Request) WarningHandler(handler WarningHandler) *Request {
```
I'm not sure. Isn't it obvious that a request only has one handler that can be set in two ways?

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -650,21 +673,16 @@ func (r *Request) tryThrottleWithInfo(ctx context.Context, retryInfo string) err
-		klog.V(3).Info(message)
+		if retryInfo == "" {
+			retryInfo = "client-side throttling, not priority and fairness"
+		}
```
`longThrottleLatency` < `extraLongThrottleLatency`, therefore we are guaranteed to enter the first if in case that any logging is taking place.

Nesting the two if statements makes that clearer and also avoids one if check in the common case that the latency is smaller than `longThrottleLatency`.

I've changed that and, while at it, also changed `var` to `const` for those two constants.

pohly commented on staging/src/k8s.io/client-go/rest/urlbackoff.go:
```diff
@@ -32,12 +34,22 @@ import (
 var serverIsOverloadedSet = sets.NewInt(429)
 var maxResponseCode = 499
 
+// Deprecated: BackoffManager.Sleep] ignores the caller's context. Use BackoffManagerWithContext instead.
```
Fixed.

pohly commented:
@soltysh: care to take another look? I added fixup commits with my changes, but I could also apply them directly if you want to do a full pass over each commit instead.

There's one small issue with out-dated go.sum files. I'll fix that during the next force-push.

pohly commented:
I rebased to align with https://github.com/kubernetes/kubernetes/pull/129125.

aojea commented:
/test pull-kubernetes-apidiff-client-go

pohly commented:
Force-pushed to test that pull-kubernetes-apidiff-client-go still gets triggered after https://github.com/kubernetes/test-infra/pull/33943 got merged.

pohly commented:
Rebased to resolve `hack/logcheck.conf` conflict after merging the client-go/tools/cache PR. In hindsight I probably shouldn't have included the `logcheck.conf` changes for each directory in those individual PRs and only do that at the end...

soltysh reviewed (commented):
A few minor nits, but the part that worries me the most is the unit test difference, currently on master I'm getting:
```
$ go test -race -cover ./staging/src/k8s.io/client-go/rest/ -count 1
ok  	k8s.io/client-go/rest	16.375s	coverage: 76.4% of statements
```
with this PR:
```
$ go test -race -cover ./staging/src/k8s.io/client-go/rest/ -count 1
ok  	k8s.io/client-go/rest	145.491s	coverage: 76.9% of statements
```
That's almost 10x more :cry:, which is a bit worrisome.

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -259,20 +259,47 @@ func (r *Request) Resource(resource string) *Request {
-// If set to nil, this client will use the default warning handler (see SetDefaultWarningHandler).
+// If set to nil, this client will use the default warning handler (see [SetDefaultWarningHandler]).
+//
+//logcheck:context // WarningHandlerWithContext should be used instead of WarningHandler in code which supports contextual logging.
```
`logcheck:context` is the final goal. It is the machine-readable annotation supported by logcheck.

The reason why we can use it here directly instead of first using `Contextual logging` and later changing to `logcheck:context` is that all code where it matters gets converted directly in this PR.

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -259,20 +259,47 @@ func (r *Request) Resource(resource string) *Request {
-// or defaults to the stub implementation if nil is provided
+// or defaults to the stub implementation if nil is provided.
+//
+// Deprecated: [BackoffManager.Sleep] ignores the caller's context. Use BackOffWithContext and [BackoffManagerWithContext] instead.
```
The intention here was to use references for easier browsing in pkg.go.dev. But even if it works (hard to test locally), then it probably would look odd in error messages from linters which show the string.

Therefore I am removing the `[]` markup again from `Deprecated:`, `Contextual Logging:` and `logcheck:contextual`.

pohly commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -129,10 +129,23 @@ type Config struct {
+	//
+	// See documentation for [SetDefaultWarningHandler] for details.
+	//
+	//logcheck:context // WarningHandlerWithContext should be used instead of WarningHandler in code which supports contextual logging.
```
https://github.com/kubernetes/kubernetes/pull/127709#discussion_r1892618868

pohly commented on staging/src/k8s.io/client-go/rest/urlbackoff.go:
```diff
@@ -49,22 +63,38 @@ type URLBackoff struct {
 	time.Sleep(d)
 }
 
+func (n *NoBackoff) SleepWithContext(ctx context.Context, d time.Duration) {
```
This is one reason why unit tests were slower.

https://github.com/kubernetes/kubernetes/blob/4e4ea1bad148cb1b0530e923c88def4a733b28fc/staging/src/k8s.io/client-go/rest/request_test.go#L2607-L2613

inherited `NoBackoff.SleepWithContext` without overriding the `SleepWithContext` method.

One could argue that this is an API break, but normally adding methods to a struct is not considered an API break and it's not flagged by apidiff as one. I think code which is meant to implement an interface should do that itself and delegate to someone else where needed, so I think the test code was at fault here. I've fixed it and most tests run quickly again.

The remaining problem is `TestBackoffLifecycle` which depends on a fake clock. The new code now uses `Clock.NewTimer` instead of `Clock.Sleep` and something there doesn't seem to fake time properly. I'm still looking.

pohly commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -1517,32 +1518,48 @@ func TestBackoffLifecycle(t *testing.T) {
 			t.Errorf("Backoff is %v instead of %v", thisBackoff, sec)
 		}
-		now := clock.Now()
-		request.DoRaw(context.Background())
```
The old code here no longer works because it relied on `Sleep` stepping forward the fake clock:
https://github.com/kubernetes/kubernetes/blob/a4b8a3b2e33a3b591884f69b64f439e6b880dc40/vendor/k8s.io/utils/clock/testing/fake_clock.go#L235-L238

Now `URLBackoff.SleepWithContext` instead calls `FakeClock.Timer` and that relies on some other goroutine to step the clock forward. That's what I am doing in the new code, but somehow it's slower. Checking...

pohly replied:
I broke it in my local dirty tree. `TestBackoffLifecycle` is okay the way it is here.

pohly replied:
What "broke" the test was this:
```patch
diff --git a/staging/src/k8s.io/client-go/rest/urlbackoff.go b/staging/src/k8s.io/client-go/rest/urlbackoff.go
index c1daa8bafa5..5b7b4e216ef 100644
--- a/staging/src/k8s.io/client-go/rest/urlbackoff.go
+++ b/staging/src/k8s.io/client-go/rest/urlbackoff.go
@@ -84,6 +84,9 @@ func (n *NoBackoff) Sleep(d time.Duration) {
 }
 
 func (n *NoBackoff) SleepWithContext(ctx context.Context, d time.Duration) {
+       if d == 0 {
+               return
+       }
        t := time.NewTimer(d)
        defer t.Stop()
        select {
@@ -147,6 +150,9 @@ func (b *URLBackoff) Sleep(d time.Duration) {
 }
 
 func (b *URLBackoff) SleepWithContext(ctx context.Context, d time.Duration) {
+       if d == 0 {
+               return
+       }
        t := b.Backoff.Clock.NewTimer(d)
        defer t.Stop()
        select {
```

This test is very sensitive to which timer functions are being called and how. It annoyed me that the test didn't handle such an optimization and also that it was less precise than before.

I had an idea how to address this: by wrapping the `SleepWithContext`, progressing the clock can be done similar to before in each sleep call. The body of the test is then exactly as before. I've made that change and also included the optimization above.

pohly commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -381,8 +394,12 @@ func RESTClientForConfigAndClient(config *Config, httpClient *http.Client) (*RES
-		restClient.warningHandler = config.WarningHandler
+	switch {
+	case err != nil:
+	case config.WarningHandlerWithContext != nil:
```
Okay.

pohly replied:
That broke `TestConfigWarningHandler/both`. I need to update the unit test to reflect the modified behavior.

pohly commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -448,8 +465,12 @@ func UnversionedRESTClientForConfigAndClient(config *Config, httpClient *http.Cl
+	case config.WarningHandlerWithContext != nil:
+		restClient.warningHandler = config.WarningHandlerWithContext
+	case config.WarningHandler != nil:
+		restClient.warningHandler = warningLoggerNopContext{l: config.WarningHandler}
```
I added `maybeSetWarningHandler`.

pohly commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -1323,16 +1346,18 @@ func (r *Request) transformResponse(ctx context.Context, resp *http.Response, re
 			// 3. Apiserver closes connection.
 			// 4. client-go should catch this and return an error.
-			klog.V(2).Infof("Stream error %#v when reading response body, may be caused by closed connection.", err)
+			klog.FromContext(ctx).V(2).Info("Stream error when reading response body, may be caused by closed connection", "err", err)
```
That would then do a `klog.FromContext` in all cases, whether there was an error or not. It's better to call it only if really needed, even if that means writing `klog.FromContext` twice.

soltysh replied:
Honestly, looking at the code it will invoke `klog.FromContext` in almost all paths, so I'm not sure I follow your argument. Also, I've noticed that `logBody` accepts context, which imo is a bit of an overkill, it should just accept logger as well. But that's not blocking this PR, can be handled some other time.

pohly replied:
You are right, I missed that this is the "normal" path. Changed.

> Also, I've noticed that logBody accepts context, which imo is a bit of an overkill

"Overkill" in the sense of "doesn't need it", not as in the sense of "performance". It's mostly for convenience: most callers don't have a logger, so if we were to require a logger, we'd end up with more `klog.FromContext` calls scattered in various places.

But if we have a logger, then passing it indeed saves some overhead for another such call. That makes sense to me, so I changed it.

> But that's not blocking this PR, can be handled some other time.

Let's do it right in this PR, such things tend to get forgotten:
https://github.com/kubernetes/kubernetes/compare/ae80977df9965a291f8634ac6b82c176f2813c88..7b6f0dfea045463da617ae00e85d3c31359c18ee

soltysh commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -1508,41 +1509,82 @@ func TestBackoffLifecycle(t *testing.T) {
+	// - SleepWithContext must be a no-op when the duration is zero
+	//   => no need to step the fake clock
+	// - SleepWithContext must use FakeClock.NewTimer, not FakeClock.Sleep
+	//   because the latter would advance time itself
```
Sounds like something that should be explained on that method as well. I mean `FakeClock.Sleep`, because I'm pretty sure other might run into similar problems, especially with these changes.

pohly replied:
`FakeClock.Sleep` documents it. It just wasn't particularly obvious that this test depends on this particular aspect.

soltysh commented on staging/src/k8s.io/client-go/rest/urlbackoff.go:
```diff
@@ -32,12 +34,24 @@ import (
 	Sleep(d time.Duration)
 }
 
+type BackoffManagerWithContext interface {
```
Nit: godoc missing.

pohly replied:
"Feature parity" with `BackoffManager`, which also doesn't have any... :shrug:

soltysh reviewed (approved):
/lgtm
/approve

/hold cancel

soltysh commented:
/lgtm

pohly commented:
/assign @liggitt 

For approval of go.mod updates and in case that you want to have a look before this merges.

pohly commented:
> A few minor nits, but the part that worries me the most is the unit test difference, currently on master I'm getting: [...] That's almost 10x more 😢, which is a bit worrisome.

For the record: this was because the fake backoff manager didn't ignore sleeps after adding `SleepWithContext`. It's fixed.

pohly commented:
Rebased because of a minor client-go/go.mod conflict.

liggitt commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,33 +51,72 @@ var (
-	klog.Warning(message)
+	// This call depth works for the "usual" call chain (Do->request->transformResponse->handleWarnings->handler).
+	// Making a guess is better than none...
+	klog.Background().WithCallDepth(8).Info("Warning", "message", message)
```
why is guessing about the depth here better than leaving this as klog.Warning(message)?

pohly replied:
Because most of the time, the guess will be right and instead of a `warnings.go: Warning <something>` app developers will get a `myniftysourcecode.go: Warning <something>` instead where `myniftysourcecode.go` is the code calling client-go.

liggitt replied:
and when the guess is wrong, it'll point at a confusing random point in the stack?

pohly replied:
Well, yes.

I don't feel strongly about this. I'll take it out.

liggitt commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,33 +51,72 @@ var (
+	}
+	// This call depth works for the "usual" call chain (Do->request->transformResponse->handleWarnings->handler).
+	// Making a guess is better than none...
+	klog.FromContext(ctx).WithCallDepth(8).Info("Warning", "message", message)
```
same question, I don't understand why guessing about the depth is a good idea

liggitt commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,33 +51,72 @@ var (
 	if code != 299 || len(message) == 0 {
 		return
 	}
-	klog.Warning(message)
```
so we no longer have a warning severity to have the warning logger use? that's frustrating

pohly replied:
Bring that up with `@thockin` over a cup of tea or something stronger... he is very passionate about not having log levels higher than info. Latest discussion: https://github.com/go-logr/logr/issues/335#issuecomment-2521775704

> logr holds as truth that "warning logs" are not a real thing. There are things worth logging and things not worth logging. Deciding at the logging call-site whether a log-line is "significant" assumes that you know that a priori. My own experience says otherwise, and the use of "warn" almost always becomes muddy over time.

pohly replied:
Done (https://github.com/kubernetes/kubernetes/compare/69101e439f9a5b9d92060e9d06534aaf960a4cf7..bce3ed3037f5fc3c85ca683d8db9c8c10717efed).

liggitt commented on staging/src/k8s.io/client-go/rest/exec_test.go:
```diff
@@ -242,6 +242,9 @@ func TestConfigToExecClusterRoundtrip(t *testing.T) {
 			*h = &fakeWarningHandler{}
 		},
+		func(h *WarningHandlerWithContext, f fuzz.Continue) {
+			*h = &fakeWarningHandlerWithContext{}
```
curious if these changes were required to pass tests or if this was just for tidyness

liggitt replied:
oh, nevermind, a new field got added

liggitt commented on staging/src/k8s.io/client-go/rest/config.go:
```diff
@@ -129,10 +129,23 @@ type Config struct {
+	// WarningHandlerWithContext handles warnings in server responses.
+	// If this and WarningHandler are not set, the
+	// default warning handler is used. If both are set,
+	// WarningHandler is used.
```
is there a reason for preferring the no-context field if both are set?

go stdlib seems to prefer the context variant if both are set. for example, from https://pkg.go.dev/net/http#Transport:

```
	// Deprecated: Use DialContext instead, which allows the transport
	// to cancel dials as soon as they are no longer needed.
	// If both are set, DialContext takes priority.

	// Deprecated: Use DialTLSContext instead, which allows the transport
	// to cancel dials as soon as they are no longer needed.
	// If both are set, DialTLSContext takes priority.
```

pohly replied:
No particular reason. I'll swap the priority.

pohly replied:
Done (https://github.com/kubernetes/kubernetes/compare/69101e439f9a5b9d92060e9d06534aaf960a4cf7..bce3ed3037f5fc3c85ca683d8db9c8c10717efed).

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -1629,7 +1660,7 @@ func (r Result) Error() error {
 	out, _, err := r.decoder.Decode(r.body, &schema.GroupVersionKind{Version: "v1"}, nil)
 	if err != nil {
-		klog.V(5).Infof("body was not decodable (unable to check for Status): %v", err)
+		r.logger.V(5).Info("Body was not decodable (unable to check for Status)", "err", err)
```
are we certain we'll never get a request here that got set up with a nil `logger`? does that need a guard?

pohly replied:
`r.logger` is not a pointer and the null struct is safe to use, therefore this [won't crash](https://github.com/go-logr/logr/blob/b8ba1898daddefb0d2a567a5b4bc420289722b9b/logr.go#L309-L311). Instead, [nothing gets logged](https://github.com/go-logr/logr/blob/b8ba1898daddefb0d2a567a5b4bc420289722b9b/logr.go#L273-L275).

To be absolutely sure that this log line gets emitted, `logger` would have to be turned into a pointer, with a fallback to `klog.Background()` if nil. But that then comes with the risk that adding a log call elsewhere forgets the if check and *that* then would crash.

There's also some additional memory overhead when using a pointer.

Overall I prefer to keep it as it is now.

liggitt replied:
sounds good, I didn't know it was a nil-safe non-pointer

pohly commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,33 +51,68 @@ var (
 	if code != 299 || len(message) == 0 {
 		return
 	}
-	klog.Warning(message)
```
This was deliberate. With structured logging, the message string is meant to be fixed. It serves as key for finding related log entries which then vary only in their additional key/value pairs.

OTOH, one can also argue that these messages are not related. They just use some common machinery for getting emitted.

I've switched to `Info("Warning: " + message)`: https://github.com/kubernetes/kubernetes/compare/bce3ed3037f5fc3c85ca683d8db9c8c10717efed..b7386467c8df686e935c477eac26049a80de789b

pohly commented on staging/src/k8s.io/client-go/rest/warnings.go:
```diff
@@ -43,33 +51,68 @@ var (
+	if code != 299 || len(message) == 0 {
+		return
+	}
+	klog.FromContext(ctx).Info("Warning", "message", message)
```
Using the second now.

liggitt commented:
/lgtm
/approve

not sure if you want to squash or not, up to you

liggitt commented:
/hold in case you want to squash

pohly commented:
No, I prefer to keep the four distinct commits. They are logically separate.

/hold cancel

Thanks for the review!
